# **Financial market News - Sentimental Analysis**

--------------------------------------------------------------------------------

# **Objective**

Financial news headlines are a rich source of NLP data, particularly for predicting stock market performance. A common approach to this task is **sentiment analysis**, which classifies text into positive, negative, or neutral sentiment categories.

In this project, the dataset consists of the top 25 financial market news headlines for the day. The objective is to train a model that can predict the overall sentiment of these headlines, helping to assess market mood and possibly guide investment decisions.

# **Import Library**

In [9]:
import pandas as pd

In [10]:
import numpy as np

# **Import dataset**

In [11]:
df = pd.read_csv('https://github.com/YBIFoundation/Dataset/raw/refs/heads/main/Financial%20Market%20News.csv', encoding= "ISO-8859-1" )

In [12]:
df.head()

Date  Label                                             News 1  \
0  01-01-2010      0           McIlroy's men catch cold from Gudjonsson   
1  02-01-2010      0               Warning from history points to crash   
2  03-01-2010      0      Comment: Why Israel's peaceniks feel betrayed   
3  04-01-2010      1             £750,000-a-goal Weah aims parting shot   
4  05-01-2010      1  Leeds arrive in Turkey to the silence of the fans   

                                       News 2  \
0                       Obituary: Brian Walsh   
1              Investors flee to dollar haven   
2  Court deals blow to seizure of drug assets   
3            Newcastle pay for Fletcher years   
4     One woman's vision offers loan lifeline   

                                            News 3  \
0       Workplace blues leave employers in the red   
1                      Banks and tobacco in favour   
2                       An ideal target for spooks   
3  Brown sent to the stands for Scotland qualifier   
4          Working Lives: How world leaders worked   

                                        News 4  \
0                     Classical review: Rattle   
1                        Review: Llama Farmers   
2  World steps between two sides intent on war   
3         Tourists wary of breaking new ground   
4           Working Lives: Tricks of the trade   

                                              News 5  \
0                     Dance review: Merce Cunningham   
1                       War jitters lead to sell-off   
2                       What the region's papers say   
3              Canary Wharf climbs into the FTSE 100   
4  Working Lives: six-hour days, long lunches and...   

                                         News 6  \
0  Genetic tests to be used in setting premiums   
1                    Your not-so-secret history   
2           Comment: Fear and rage in Palestine   
3                           Review: Bill Bailey   
4                        Pop review: We Love UK   

                                          News 7  \
0                        Opera review: La Bohème   
1                  Review: The Northern Sinfonia   
2  Poverty and resentment fuels Palestinian fury   
3                              Review: Classical   
4               World music review: Marisa Monte   

                                        News 8  ...  \
0                   Pop review: Britney Spears  ...   
1                             Review: Hysteria  ...   
2  Republican feud fear as dissident is killed  ...   
3              Review: New Contemporaries 2000  ...   
4             Art review:  Hollingsworth/Heyer  ...   

                                News 16                              News 17  \
0                 Finland 0 - 0 England                   Healy a marked man   
1  Why Wenger will stick to his Gunners  Out of luck England hit rock bottom   
2             FTSE goes upwardly mobile             At this price?  BP Amoco   
3         More cash on way for counties      Cairns carries Kiwis to victory   
4          Duisenberg in double trouble           Pru to cut pension charges   

                                             News 18  \
0                     Happy birthday Harpers & Queen   
1                         Wilkinson out of his depth   
2                                            Go fish   
3  Year of Blanchflower's flourish when Spurs sto...   
4                            Art review: Paul Graham   

                                             News 19  \
0  Win unlimited access to the Raindance film fes...   
1                   Kinsella sparks Irish power play   
2         Bosnian Serb blows himself up to evade law   
3      New direct approach brings only pay-per-blues   
4               Shearer shot sparks Boro humiliation   

                                             News 20  \
0  Labour pledges £800m to bridge north-south divide   
1                    Brown banished as Scots rebound   
2                       Or

In [13]:
df.tail()

Date  Label                                             News 1  \
4096  20-03-2021      0  Barclays and RBS shares suspended from trading...   
4097  21-03-2021      1  2,500 Scientists To Australia: If You Want To ...   
4098  22-03-2021      1                   Explosion At Airport In Istanbul   
4099  23-03-2021      1  Jamaica proposes marijuana dispensers for tour...   
4100  24-03-2021      1  A 117-year-old woman in Mexico City finally re...   

                                                 News 2  \
4096  Pope says Church should ask forgiveness from g...   
4097  The personal details of 112,000 French police ...   
4098  Yemeni former president: Terrorism is the offs...   
4099  Stephen Hawking says pollution and 'stupidity'...   
4100   IMF chief backs Athens as permanent Olympic host   

                                                 News 3  \
4096  Poland 'shocked' by xenophobic abuse of Poles ...   
4097  S&amp;P cuts United Kingdom sovereign credit r...   
4098  UK must accept freedom of movement to access E...   
4099  Boris Johnson says he will not run for Tory pa...   
4100  The president of France says if Brexit won, so...   

                                                 News 4  \
4096  There will be no second referendum, cabinet ag...   
4097                Huge helium deposit found in Africa   
4098  Devastated: scientists too late to captive bre...   
4099  Six gay men in Ivory Coast were abused and for...   
4100  British Man Who Must Give Police 24 Hours' Not...   

                                                 News 5  \
4096      Scotland welcome to join EU, Merkel ally says   
4097  CEO of the South African state broadcaster qui...   
4098  British Labor Party leader Jeremy Corbyn loses...   
4099  Switzerland denies citizenship to Muslim immig...   
4100  100+ Nobel laureates urge Greenpeace to stop o...   

                                                 News 6  \
4096  Sterling dips below Friday's 31-year low amid ...   
4097  Brexit cost investors $2 trillion, the worst o...   
4098  A Muslim Shop in the UK Was Just Firebombed Wh...   
4099  Palestinian terrorist stabs israeli teen girl ...   
4100  Brazil: Huge spike in number of police killing...   

                                                 News 7  \
4096  No negative news about South African President...   
4097  Hong Kong democracy activists call for return ...   
4098  Mexican Authorities Sexually Torture Women in ...   
4099  Puerto Rico will default on $1 billion of debt...   
4100  Austria's highest court annuls presidential el...   

                                                 News 8  ...  \
4096  Surge in Hate Crimes in the U.K. Following U.K...  ...   
4097  Brexit: Iceland president says UK can join 'tr...  ...   
4098            UK shares and pound continue to recover  ...   
4099  Republic of Ireland fans to be awarded medal f...  ...   
4100  Facebook wins privacy case, can track any Belg...  ...   

                                                News 16  \
4096  German lawyers to probe Erdogan over alleged w...   
4097  US, Canada and Mexico pledge 50% of power from...   
4098  Escape Tunnel, Dug by Hand, Is Found at Holoca...   
4099  Googles free wifi at Indian railway stations i...   
4100  The United States has placed Myanmar, Uzbekist...   

                                                News 17  \
4096  Boris Johnson says the UK will continue to "in...   
4097  There is increasing evidence that Australia is...   
4098  The land under Beijing is sinking by as much a...   
4099  Mounting evidence suggests 'hobbits' were wipe...   
4100  S&amp;P revises European Union credit rating t...   

                                                News 18  \
4096  Richard Branson is calling on the UK governmen...   
4097  Richard Branson, the founder of Virgin Group, ...   
4098  Car bomb and Anti-Islamic attack on Mosque in ...   
4099  The men who carried out Tuesday's terror attac...   
4100  India gets $1 bil

# **Describe Data**

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4101 entries, 0 to 4100
Data columns (total 27 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Date     4101 non-null   object
 1   Label    4101 non-null   int64 
 2   News 1   4101 non-null   object
 3   News 2   4101 non-null   object
 4   News 3   4101 non-null   object
 5   News 4   4101 non-null   object
 6   News 5   4101 non-null   object
 7   News 6   4101 non-null   object
 8   News 7   4101 non-null   object
 9   News 8   4101 non-null   object
 10  News 9   4101 non-null   object
 11  News 10  4101 non-null   object
 12  News 11  4101 non-null   object
 13  News 12  4101 non-null   object
 14  News 13  4101 non-null   object
 15  News 14  4101 non-null   object
 16  News 15  4101 non-null   object
 17  News 16  4101 non-null   object
 18  News 17  4101 non-null   object
 19  News 18  4101 non-null   object
 20  News 19  4101 non-null   object
 21  News 20  4101 non-null   object
 22  

In [15]:
df.describe()

Label
count  4101.000000
mean      0.528164
std       0.499267
min       0.000000
25%       0.000000
50%       1.000000
75%       1.000000
max       1.000000

In [16]:
df.shape

(4101, 27)

In [17]:
df.columns

Index(['Date', 'Label', 'News 1', 'News 2', 'News 3', 'News 4', 'News 5',
       'News 6', 'News 7', 'News 8', 'News 9', 'News 10', 'News 11', 'News 12',
       'News 13', 'News 14', 'News 15', 'News 16', 'News 17', 'News 18',
       'News 19', 'News 20', 'News 21', 'News 22', 'News 23', 'News 24',
       'News 25'],
      dtype='object')

# **Feature Selection**

In [18]:
' '.join(str(x) for x in df.iloc[1,2:27])

"Warning from history points to crash Investors flee to dollar haven Banks and tobacco in favour Review: Llama Farmers War jitters lead to sell-off Your not-so-secret history Review: The Northern Sinfonia Review: Hysteria Review: The Guardsman Opera: The Marriage of Figaro Review: The Turk in Italy Deutsche spells out its plans for diversification Traders' panic sends oil prices skyward TV sport chief leaves home over romance Leader: Hi-tech twitch Why Wenger will stick to his Gunners Out of luck England hit rock bottom Wilkinson out of his depth Kinsella sparks Irish power play Brown banished as Scots rebound Battling Wales cling to lifeline Ehiogu close to sealing Boro move Man-to-man marking Match stats French referee at centre of storm is no stranger to controversy"

In [19]:
df.index

RangeIndex(start=0, stop=4101, step=1)

In [20]:
len(df.index)

4101

In [21]:
news = []
for row in range(0, len(df.index)):
    news.append(' '.join(str(x) for x in df.iloc[row, 2:27]))

In [22]:
type(news)

list

In [23]:
news[0]

"McIlroy's men catch cold from Gudjonsson Obituary: Brian Walsh Workplace blues leave employers in the red Classical review: Rattle Dance review: Merce Cunningham Genetic tests to be used in setting premiums Opera review: La Bohème Pop review: Britney Spears Theatre review: The Circle Wales face a fraught night Under-21  round-up Smith off to blot his copybook Finns taking the mickey Praise wasted as Brown studies injury options Ireland wary of minnows Finland 0 - 0 England Healy a marked man Happy birthday Harpers & Queen Win unlimited access to the Raindance film festival Labour pledges £800m to bridge north-south divide Wales: Lib-Lab pact firm despite resignation Donald Dewar Regenerating homes  regenerates well-being in people Win £100 worth of underwear TV guide: Random views"

In [24]:
X = news

In [25]:
type(X)

list

# **Text conversion to bag of words**

In [26]:
from sklearn.feature_extraction.text import CountVectorizer

In [27]:
cv = CountVectorizer(lowercase = True, ngram_range = (1,1))

In [28]:
X = cv.fit_transform(X)

In [29]:
X.shape

(4101, 48527)

In [30]:
y = df['Label']

In [31]:
y.shape

(4101,)

# **Train Test Split dataset**

In [32]:
from sklearn.model_selection import train_test_split

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, stratify = y, random_state = 2520)

In [34]:
from sklearn.ensemble import RandomForestClassifier

In [35]:
rf = RandomForestClassifier(n_estimators = 200)

In [36]:
rf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=200)

In [37]:
y_pred = rf.predict(X_test)

In [38]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [39]:
confusion_matrix(y_test, y_pred)

array([[130, 451],
       [153, 497]])

In [40]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.46      0.22      0.30       581
           1       0.52      0.76      0.62       650

    accuracy                           0.51      1231
   macro avg       0.49      0.49      0.46      1231
weighted avg       0.49      0.51      0.47      1231



# **Explanation**

This result is based on a binary classification task, where there are two classes: 0 and 1.
 Let's break down the key metrics:

This result is based on a binary classification task, where the model predicts one of two classes: 0 or 1. Let’s break down the key performance metrics:

**Precision:**

Precision measures how many of the predicted positives were actually correct. For class 0, the precision is 0.46, meaning 46% of the instances predicted as class 0 were indeed class 0. For class 1, the precision is 0.52, indicating 52% of the instances predicted as class 1 were true positives.

**Recall (Sensitivity):**

Recall, or True Positive Rate, measures how well the model identifies actual positives. For class 0, the recall is 0.22, indicating that only 22% of the true class 0 instances were correctly predicted. For class 1, recall is higher at 0.77, meaning 77% of the true class 1 instances were correctly identified.

**F1-Score:**

The F1-score is the harmonic mean of precision and recall, offering a balanced metric when you want to consider both false positives and false negatives. For class 0, the F1-score is 0.29, and for class 1, it’s 0.62. A higher F1-score indicates a better balance between precision and recall.

**Support:**

Support reflects the number of instances in each class. Here, there are 581 instances of class 0 and 650 instances of class 1 in the dataset.

**Accuracy:**

Accuracy measures the overall correctness of the model by calculating the proportion of correct predictions out of all predictions. In this case, the model’s overall accuracy is 51%.

**Macro Average & Weighted Average:**

- The **macro average** is the unweighted average of precision, recall, and F1-score across both classes, treating them equally.
- The **weighted average** takes into account the number of instances in each class, giving more weight to classes with higher support. This provides a more representative performance measure when class distribution is imbalanced.
